drivers/shared/executor/exec_utils.go

```diff
@@ -0,0 +1,295 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+
+	hclog "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/plugins/drivers"
+	dproto "github.com/hashicorp/nomad/plugins/drivers/proto"
+)
+
+// execHelper is a convenient wrapper for starting and executing commands, and handling their output
+type execHelper struct {
+	logger hclog.Logger
+
+	// newTerminal function creates a tty appropriate for the command
+	// The returned master end of tty function is to be called after process start.
+	newTerminal func() (master func() (*os.File, error), slave *os.File, err error)
```

Comment:
Let's use parent / child terminology instead like https://godoc.org/github.com/opencontainers/runc/libcontainer/utils#NewSockPair

Reply:
This is a bit strange of a pattern as well. What if you made all of these funcs methods on an interface like:

```go
type terminal interface{
  newTerminal() (...)
  setTTY(...) error
  ...
}
```

...and then created structs that implemented them? Might be a little more code, but I think it might be more idiomatic and unittestable?

Reply:
It's a bit confusing - libcontainer uses Parent/child for the socket that's controlling the tty - but the tty documentation uses master/slave e.g. https://linux.die.net/man/4/pts .

libcontainer avoids using master/slave by using RecvFd[1] and SendFd[2] but they are going against pretty much all tty and tty libraries documentation.  Recognize the history of terminology, but I would rather not confuse future devs who need to consult publicly available documentations for tty.

[1] https://godoc.org/github.com/opencontainers/runc/libcontainer/utils#RecvFd
[2] https://godoc.org/github.com/opencontainers/runc/libcontainer/utils#SendFd

Reply:
Let's break from the historical terminology on this one and use libcontainer's terminology. The kr/pty recently fixed their terminology as well: https://github.com/kr/pty/commit/7dc38fb350b1d71383eed149e73acb7bae231ddb

Reply:
Umm.. actually looks like libcontainer still calls them master/slave in the docs[1] and in its implementation [2], and so does containerd/console[3] (the library used by runc to create console and created by ~the same maintainers) when referring to the console fields.  I'll use kr/pty terminology instead.

[1] https://github.com/opencontainers/runc/blob/eb4aeed24ffbf8e2d740fafea39d91faa0ee84d0/docs/terminals.md#-new-terminal
[2] https://github.com/opencontainers/runc/blob/master/libcontainer/console_linux.go#L9-L41
[3] https://github.com/containerd/console/blob/0650fd9eeb50bab4fc99dceb9f2e14cf58f36e7f/console_unix.go#L27-L30